File: 3.无重复字符的最长子串.cc
#include <string>
#include <vector>
#include <algorithm>
using std::string;
using std::vector;
using std::max;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> chars(128);
        int left = 0, right = 0;
        int ans = 0;
        
        while (right < s.length()) {
            char r = s[right];
            chars[r]++;
            
            while (chars[r] > 1) {
                char l = s[left];
                chars[l]--;
                left++;
            }
            
            ans = max(ans, right - left + 1);
            right++;
        }
        
        return ans;
    }
};

File: Python/3.无重复字符的最长子串.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans: int = 0
        if len(s) == 0:
            return ans
        arr: list[bool] = [False] * 128
        left, right = 0, 0
        while right < len(s):
            if arr[ord(s[right])] == False:
                arr[ord(s[right])] = True
                ans = max(ans, right - left + 1)
                right += 1
            else:
                arr[ord(s[left])] = False
                left += 1
        return ans
